File: src/test_qjson.py
#!/usr/bin/env python3
# ============================================================
# test_qjson.py — Tests for QJSON parser/serializer
# ============================================================

import os, sys
sys.path.insert(0, os.path.dirname(__file__))
from decimal import Decimal
from qjson import parse, stringify, BigInt, BigFloat, Blob, js64_encode, js64_decode

passed = 0
failed = 0

def test(name, fn):
    global passed, failed
    try:
        fn()
        passed += 1
        print("  \u2713 " + name)
    except Exception as e:
        failed += 1
        print("  \u2717 " + name + ": " + str(e))


# ── Backward compat: valid JSON parses correctly ─────────────

def test_json_compat():
    assert parse('42') == 42
    assert parse('-3.14') == -3.14
    assert parse('"hello"') == "hello"
    assert parse('true') is True
    assert parse('false') is False
    assert parse('null') is None
    assert parse('{"a":1,"b":[2,3]}') == {"a": 1, "b": [2, 3]}
    assert parse('[]') == []
    assert parse('{}') == {}

def test_string_escapes():
    assert parse(r'"hello\nworld"') == "hello\nworld"
    assert parse(r'"tab\there"') == "tab\there"
    assert parse(r'"quote\""') == 'quote"'
    assert parse(r'"slash\\"') == "slash\\"
    assert parse(r'"unicode\u0041"') == "unicodeA"

# ── Comments ─────────────────────────────────────────────────

def test_line_comments():
    assert parse('// leading\n42') == 42
    assert parse('42 // trailing') == 42
    assert parse('{\n// comment\n"a": 1\n}') == {"a": 1}

def test_block_comments():
    assert parse('/* before */ 42') == 42
    assert parse('42 /* after */') == 42
    assert parse('{"a": /* inline */ 1}') == {"a": 1}

def test_mixed_comments():
    text = """
    {
      // line comment
      "x": 1,
      /* block
         comment */
      "y": 2
    }
    """
    assert parse(text) == {"x": 1, "y": 2}

def test_nested_block_comments():
    assert parse('/* outer /* inner */ still comment */ 42') == 42
    text = '{ /* a /* b */ c */ "x": 1 }'
    assert parse(text) == {"x": 1}

# ── Trailing commas ──────────────────────────────────────────

def test_trailing_comma_object():
    assert parse('{"a": 1, "b": 2,}') == {"a": 1, "b": 2}

def test_trailing_comma_array():
    assert parse('[1, 2, 3,]') == [1, 2, 3]

def test_trailing_comma_nested():
    assert parse('{"a": [1, 2,], "b": {"c": 3,},}') == {"a": [1, 2], "b": {"c": 3}}

# ── Unquoted keys ────────────────────────────────────────────

def test_unquoted_keys():
    assert parse('{name: "alice", age: 30}') == {"name": "alice", "age": 30}

def test_unquoted_keys_underscore():
    assert parse('{_id: 1, $ref: "x"}') == {"_id": 1, "$ref": "x"}

def test_mixed_keys():
    assert parse('{"quoted": 1, bare: 2}') == {"quoted": 1, "bare": 2}

# ── BigInt ───────────────────────────────────────────────────

def test_bigint_parse():
    v = parse('42N')
    assert isinstance(v, BigInt)
    assert v == 42

def test_bigint_large():
    v = parse('98765432101234567890N')
    assert isinstance(v, BigInt)
    assert v == 98765432101234567890

def test_bigint_negative():
    v = parse('-123N')
    assert isinstance(v, BigInt)
    assert v == -123

def test_bigint_in_object():
    v = parse('{"nonce": 42n, "name": "test"}')
    assert isinstance(v["nonce"], BigInt)
    assert v["nonce"] == 42
    assert v["name"] == "test"

# ── BigDecimal ───────────────────────────────────────────────

def test_bigdecimal_parse():
    v = parse('3.14M')
    assert isinstance(v, Decimal)
    assert v == Decimal("3.14")

def test_bigdecimal_integer():
    v = parse('100M')
    assert isinstance(v, Decimal)
    assert v == Decimal("100")

def test_bigdecimal_precision():
    v = parse('3.141592653589793238462643383279M')
    assert isinstance(v, Decimal)
    assert str(v) == '3.141592653589793238462643383279'

def test_bigdecimal_negative():
    v = parse('-0.001M')
    assert isinstance(v, Decimal)
    assert v == Decimal("-0.001")

# ── BigFloat ─────────────────────────────────────────────────

def test_bigfloat_parse():
    v = parse('3.14L')
    assert isinstance(v, BigFloat)
    assert float(v) == 3.14

def test_bigfloat_precision():
    v = parse('3.141592653589793238462643383279L')
    assert isinstance(v, BigFloat)
    assert str(v) == '3.141592653589793238462643383279'

def test_bigfloat_negative():
    v = parse('-1.5L')
    assert isinstance(v, BigFloat)
    assert float(v) == -1.5

def test_bigfloat_integer_form():
    v = parse('100L')
    assert isinstance(v, BigFloat)
    assert float(v) == 100.0

def test_bigint_lowercase_accepted():
    v = parse('99n')
    assert isinstance(v, BigInt)
    assert stringify(v) == '99N'

def test_bigdecimal_lowercase_accepted():
    v = parse('1.5m')
    assert isinstance(v, Decimal)
    assert stringify(v) == '1.5M'

def test_bigfloat_lowercase_accepted():
    v = parse('2.718l')
    assert isinstance(v, BigFloat)
    assert stringify(v) == '2.718L'

# ── Serializer ───────────────────────────────────────────────

def test_stringify_basic():
    assert stringify(42) == '42'
    assert stringify("hi") == '"hi"'
    assert stringify(True) == 'true'
    assert stringify(None) == 'null'
    assert stringify([1, 2]) == '[1,2]'
    assert stringify({"a": 1}) == '{"a":1}'

def test_stringify_bigint():
    assert stringify(BigInt(42)) == '42N'
    assert stringify(BigInt(-99)) == '-99N'
    assert stringify(BigInt(98765432101234567890)) == '98765432101234567890N'

def test_stringify_bigdecimal():
    assert stringify(Decimal("3.14")) == '3.14M'
    assert stringify(Decimal("-0.001")) == '-0.001M'

def test_stringify_bigfloat():
    assert stringify(BigFloat("3.14")) == '3.14L'
    assert stringify(BigFloat("3.141592653589793238462643383279")) == '3.141592653589793238462643383279L'

def test_stringify_nested():
    obj = {"amount": Decimal("99.99"), "count": BigInt(7), "tags": ["a", "b"]}
    s = stringify(obj)
    assert '"amount":99.99M' in s
    assert '"count":7N' in s

# ── Round-trip ───────────────────────────────────────────────

def test_roundtrip_bigint():
    v = BigInt(12345678901234567890)
    assert parse(stringify(v)) == v

def test_roundtrip_bigdecimal():
    v = Decimal("3.141592653589793238462643383279")
    assert parse(stringify(v)) == v

def test_roundtrip_bigfloat():
    v = BigFloat("3.141592653589793238462643383279")
    assert parse(stringify(v)) == v

def test_roundtrip_complex():
    obj = {
        "n": BigInt(42),
        "d": Decimal("1.5"),
        "s": "hello",
        "a": [1, BigInt(2), Decimal("3.0")],
        "nested": {"ok": True}
    }
    assert parse(stringify(obj)) == obj

def test_roundtrip_regular_json():
    obj = {"a": 1, "b": [2, 3.5, "x", None, True, False]}
    assert parse(stringify(obj)) == obj


# ── Run ──────────────────────────────────────────────────────

print("qjson.py")
test("JSON backward compat", test_json_compat)
test("string escapes", test_string_escapes)
test("line comments", test_line_comments)
test("block comments", test_block_comments)
test("mixed comments", test_mixed_comments)
test("nested block comments", test_nested_block_comments)
test("trailing comma object", test_trailing_comma_object)
test("trailing comma array", test_trailing_comma_array)
test("trailing comma nested", test_trailing_comma_nested)
test("unquoted keys", test_unquoted_keys)
test("unquoted keys _ and $", test_unquoted_keys_underscore)
test("mixed quoted/bare keys", test_mixed_keys)
test("BigInt parse", test_bigint_parse)
test("BigInt large", test_bigint_large)
test("BigInt negative", test_bigint_negative)
test("BigInt in object", test_bigint_in_object)
test("BigDecimal parse", test_bigdecimal_parse)
test("BigDecimal integer form", test_bigdecimal_integer)
test("BigDecimal precision", test_bigdecimal_precision)
test("BigDecimal negative", test_bigdecimal_negative)
test("BigFloat parse", test_bigfloat_parse)
test("BigFloat precision", test_bigfloat_precision)
test("BigFloat negative", test_bigfloat_negative)
test("BigFloat integer form", test_bigfloat_integer_form)
test("BigInt lowercase n accepted", test_bigint_lowercase_accepted)
test("BigDecimal lowercase m accepted", test_bigdecimal_lowercase_accepted)
test("BigFloat lowercase l accepted", test_bigfloat_lowercase_accepted)
test("stringify basic", test_stringify_basic)
test("stringify BigInt", test_stringify_bigint)
test("stringify BigDecimal", test_stringify_bigdecimal)
test("stringify BigFloat", test_stringify_bigfloat)
test("stringify nested", test_stringify_nested)
test("round-trip BigInt", test_roundtrip_bigint)
test("round-trip BigDecimal", test_roundtrip_bigdecimal)
test("round-trip BigFloat", test_roundtrip_bigfloat)
test("round-trip complex", test_roundtrip_complex)
test("round-trip regular JSON", test_roundtrip_regular_json)

# ── Blob / JS64 tests ──────────────────────────────────────

def test_js64_roundtrip():
    hello = b"\x48\x65\x6c\x6c\x6f"  # "Hello"
    enc = js64_encode(hello)
    dec = js64_decode(enc)
    assert dec == hello, "round-trip Hello"

def test_js64_empty():
    enc = js64_encode(b"")
    assert enc == ""
    dec = js64_decode("")
    assert dec == b""

def test_js64_single_byte():
    enc = js64_encode(b"\xff")
    dec = js64_decode(enc)
    assert dec == b"\xff"

def test_blob_parse():
    hello = b"\x48\x65\x6c\x6c\x6f"
    enc = js64_encode(hello)
    obj = parse("0j" + enc)
    assert isinstance(obj, Blob)
    assert obj.data == hello

def test_blob_parse_uppercase():
    enc = js64_encode(b"\x48\x65")
    obj = parse("0J" + enc)
    assert isinstance(obj, Blob)
    assert obj.data == b"\x48\x65"

def test_blob_in_object():
    enc = js64_encode(b"\x01\x02\x03")
    obj = parse("{key: 0j" + enc + "}")
    assert isinstance(obj["key"], Blob)
    assert obj["key"].data == b"\x01\x02\x03"

def test_blob_stringify_roundtrip():
    hello = b"\x48\x65\x6c\x6c\x6f"
    text = stringify(Blob(hello))
    assert text.startswith("0j"), "starts with 0j"
    rt = parse(text)
    assert isinstance(rt, Blob)
    assert rt.data == hello

def test_blob_empty():
    obj = parse("0j")
    assert isinstance(obj, Blob)
    assert obj.data == b""

test("JS64 round-trip", test_js64_roundtrip)
test("JS64 empty", test_js64_empty)
test("JS64 single byte", test_js64_single_byte)
test("blob parse", test_blob_parse)
test("blob parse uppercase", test_blob_parse_uppercase)
test("blob in object", test_blob_in_object)
test("blob stringify round-trip", test_blob_stringify_roundtrip)
test("blob empty", test_blob_empty)

print("\n%d tests: %d passed, %d failed" % (passed + failed, passed, failed))
if failed:
    sys.exit(1)

File: src/test-qjson.js
// ============================================================
// test-qjson.js — Tests for QJSON parser/serializer
//
// Run:  node src/test-qjson.js
//       bun run src/test-qjson.js
// ============================================================

import { qjson_parse, qjson_stringify, js64_encode, js64_decode } from './qjson.js';

var passed = 0, failed = 0;
var hasBigInt = typeof BigInt !== "undefined";

function test(name, fn) {
  try { fn(); passed++; console.log("  \u2713 " + name); }
  catch(e) { failed++; console.log("  \u2717 " + name + ": " + e.message); }
}
function assert(cond, msg) { if (!cond) throw new Error(msg || "assertion failed"); }
function eq(a, b, msg) { assert(a === b, msg || "expected " + b + ", got " + a); }

// ── JSON backward compat ────────────────────────────────────

console.log("qjson.js");

test("JSON compat: primitives", function() {
  eq(qjson_parse("42"), 42);
  eq(qjson_parse("-3.14"), -3.14);
  eq(qjson_parse('"hello"'), "hello");
  eq(qjson_parse("true"), true);
  eq(qjson_parse("false"), false);
  eq(qjson_parse("null"), null);
});

test("JSON compat: objects and arrays", function() {
  var obj = qjson_parse('{"a":1,"b":[2,3]}');
  eq(obj.a, 1);
  eq(obj.b.length, 2);
  eq(obj.b[0], 2);
  var arr = qjson_parse("[]");
  eq(arr.length, 0);
});

test("string escapes", function() {
  eq(qjson_parse('"hello\\nworld"'), "hello\nworld");
  eq(qjson_parse('"tab\\there"'), "tab\there");
  eq(qjson_parse('"quote\\""'), 'quote"');
  eq(qjson_parse('"unicode\\u0041"'), "unicodeA");
});

// ── Comments ────────────────────────────────────────────────

test("line comments", function() {
  eq(qjson_parse("// leading\n42"), 42);
  eq(qjson_parse("42 // trailing"), 42);
  var obj = qjson_parse('{\n// comment\n"a": 1\n}');
  eq(obj.a, 1);
});

test("block comments", function() {
  eq(qjson_parse("/* before */ 42"), 42);
  eq(qjson_parse("42 /* after */"), 42);
  var obj = qjson_parse('{"a": /* inline */ 1}');
  eq(obj.a, 1);
});

test("mixed comments", function() {
  var obj = qjson_parse('{\n  // line\n  "x": 1,\n  /* block */\n  "y": 2\n}');
  eq(obj.x, 1);
  eq(obj.y, 2);
});

test("nested block comments", function() {
  eq(qjson_parse("/* outer /* inner */ still */ 42"), 42);
});

test("trailing commas", function() {
  var obj = qjson_parse('{"a": 1, "b": 2,}');
  eq(obj.a, 1); eq(obj.b, 2);
  var arr = qjson_parse("[1, 2, 3,]");
  eq(arr.length, 3); eq(arr[2], 3);
});

test("unquoted keys", function() {
  var obj = qjson_parse('{name: "alice", age: 30}');
  eq(obj.name, "alice"); eq(obj.age, 30);
});

test("unquoted keys _ and $", function() {
  var obj = qjson_parse('{_id: 1, $ref: "x"}');
  eq(obj._id, 1); eq(obj.$ref, "x");
});

test("mixed quoted/bare keys", function() {
  var obj = qjson_parse('{"quoted": 1, bare: 2}');
  eq(obj.quoted, 1); eq(obj.bare, 2);
});

// ── BigInt ──────────────────────────────────────────────────

test("BigInt parse", function() {
  var v = qjson_parse("42N");
  if (hasBigInt) {
    eq(typeof v, "bigint");
    assert(v === BigInt(42), "expected 42n");
  } else {
    eq(v, "42N");  // fallback: string
  }
});

test("BigInt negative", function() {
  var v = qjson_parse("-123N");
  if (hasBigInt) {
    assert(v === BigInt(-123), "expected -123n");
  }
});

test("BigInt in object", function() {
  var obj = qjson_parse('{"nonce": 42n, "name": "test"}');
  eq(obj.name, "test");
  if (hasBigInt) {
    eq(typeof obj.nonce, "bigint");
  }
});

// ── BigDecimal ──────────────────────────────────────────────

test("BigDecimal parse", function() {
  var v = qjson_parse("3.14M");
  // On QuickJS: bigdecimal. On Node: string fallback.
  if (typeof BigDecimal !== "undefined") {
    eq(typeof v, "bigdecimal");
  } else {
    eq(v, "3.14M");
  }
});

test("BigDecimal integer form", function() {
  var v = qjson_parse("100M");
  if (typeof BigDecimal !== "undefined") {
    eq(typeof v, "bigdecimal");
  } else {
    eq(v, "100M");
  }
});

// ── BigFloat ────────────────────────────────────────────────

test("BigFloat parse", function() {
  var v = qjson_parse("3.14L");
  // On QuickJS: bigfloat. On Node/Bun: string fallback.
  if (typeof BigFloat !== "undefined") {
    eq(typeof v, "bigfloat");
  } else {
    eq(v, "3.14L");
  }
});

test("BigFloat precision preserved", function() {
  var v = qjson_parse("3.141592653589793238462643383279L");
  if (typeof BigFloat !== "undefined") {
    eq(typeof v, "bigfloat");
  } else {
    eq(v, "3.141592653589793238462643383279L");
  }
});

test("lowercase n/m/l all accepted", function() {
  // BigInt lowercase
  var bi = qjson_parse("99n");
  if (hasBigInt) {
    assert(bi === BigInt(99), "99n should parse as BigInt");
  } else {
    eq(bi, "99N");
  }
  // BigDecimal lowercase
  var bd = qjson_parse("1.5m");
  if (typeof BigDecimal !== "undefined") {
    eq(typeof bd, "bigdecimal");
  } else {
    eq(bd, "1.5M");
  }
  // BigFloat lowercase
  var bf = qjson_parse("2.718l");
  if (typeof BigFloat !== "undefined") {
    eq(typeof bf, "bigfloat");
  } else {
    eq(bf, "2.718L");
  }
});

// ── Serializer ──────────────────────────────────────────────

test("stringify basic", function() {
  eq(qjson_stringify(42), "42");
  eq(qjson_stringify("hi"), '"hi"');
  eq(qjson_stringify(true), "true");
  eq(qjson_stringify(null), "null");
  eq(qjson_stringify([1, 2]), "[1,2]");
  eq(qjson_stringify({a: 1}), '{"a":1}');
});

test("stringify BigInt", function() {
  if (!hasBigInt) return;  // skip on runtimes without BigInt
  eq(qjson_stringify(BigInt(42)), "42N");
  eq(qjson_stringify(BigInt(-99)), "-99N");
});

test("stringify nested", function() {
  if (!hasBigInt) return;
  var s = qjson_stringify({count: BigInt(7), tags: ["a"]});
  assert(s.indexOf("7N") > -1, "should contain 7N");
  assert(s.indexOf('"a"') > -1, "should contain tag");
});

// ── Round-trip ──────────────────────────────────────────────

test("round-trip BigInt", function() {
  if (!hasBigInt) return;
  var v = BigInt(12345678901234567890);
  var rt = qjson_parse(qjson_stringify(v));
  assert(rt === v, "BigInt round-trip failed");
});

test("round-trip regular JSON", function() {
  var obj = {a: 1, b: [2, 3.5, "x", null, true, false]};
  var rt = qjson_parse(qjson_stringify(obj));
  eq(rt.a, 1);
  eq(rt.b[1], 3.5);
  eq(rt.b[2], "x");
  eq(rt.b[3], null);
});

test("round-trip complex", function() {
  if (!hasBigInt) return;
  var obj = {n: BigInt(42), s: "hello", a: [1, BigInt(2)], nested: {ok: true}};
  var rt = qjson_parse(qjson_stringify(obj));
  assert(rt.n === BigInt(42), "n should be 42n");
  eq(rt.s, "hello");
  assert(rt.a[1] === BigInt(2), "a[1] should be 2n");
  eq(rt.nested.ok, true);
});

// ── Blob / JS64 tests ──────────────────────────────────────

test("JS64 encode/decode round-trip", function() {
  var hello = [0x48, 0x65, 0x6c, 0x6c, 0x6f]; // "Hello"
  var enc = js64_encode(hello);
  var dec = js64_decode(enc);
  eq(dec.length, 5);
  eq(dec[0], 0x48);
  eq(dec[4], 0x6f);
});

test("JS64 empty round-trip", function() {
  var enc = js64_encode([]);
  eq(enc, "");
  var dec = js64_decode("");
  eq(dec.length, 0);
});

test("JS64 single byte 0xFF", function() {
  var enc = js64_encode([0xFF]);
  var dec = js64_decode(enc);
  eq(dec.length, 1);
  eq(dec[0], 0xFF);
});

test("blob parse 0j", function() {
  var hello = [0x48, 0x65, 0x6c, 0x6c, 0x6f];
  var enc = js64_encode(hello);
  var obj = qjson_parse("0j" + enc);
  eq(obj.$qjson, "blob");
  eq(obj.data.length, 5);
  eq(obj.data[0], 0x48);
  eq(obj.data[4], 0x6f);
});

test("blob parse 0J (uppercase)", function() {
  var enc = js64_encode([0x48, 0x65]);
  var obj = qjson_parse("0J" + enc);
  eq(obj.$qjson, "blob");
  eq(obj.data.length, 2);
});

test("blob in object", function() {
  var enc = js64_encode([1, 2, 3]);
  var obj = qjson_parse("{key: 0j" + enc + "}");
  eq(obj.key.$qjson, "blob");
  eq(obj.key.data.length, 3);
  eq(obj.key.data[0], 1);
});

test("blob stringify round-trip", function() {
  var hello = [0x48, 0x65, 0x6c, 0x6c, 0x6f];
  var obj = { $qjson: "blob", data: hello };
  var text = qjson_stringify(obj);
  assert(text.indexOf("0j") === 0, "starts with 0j");
  var rt = qjson_parse(text);
  eq(rt.$qjson, "blob");
  eq(rt.data.length, 5);
  eq(rt.data[0], 0x48);
  eq(rt.data[4], 0x6f);
});

test("empty blob parse", function() {
  var obj = qjson_parse("0j");
  eq(obj.$qjson, "blob");
  eq(obj.data.length, 0);
});

console.log("\n" + (passed + failed) + " tests: " + passed + " passed, " + failed + " failed");
if (failed) process.exit(1);
